Engineering thesis/rozdzialy/Implementacja.tex

```diff
@@ -0,0 +1,727 @@
+                increment = -1 * threshold
+            } else if (pulseWidth <= 1000) {
+                increment = threshold
+            }
+        }, threshold)
+        setTimeout(() => {
+            clearInterval(servoInterval)
+            resolve()
+        }, 20000)
+    })}
+\end{lstlisting}
+W celu spełnienia założeń przy sterowaniu serwomechanizmem zaimplementowano funkcję. Wykorzystuje ona technikę sterowania sygnałami PWM. Przy zastosowaniu interwałów czasowych, szerokość impulsów zwiększa się co określoną jednostkę czasową. Jeżeli serwomechanizm osiągnie pozycję końcową następuje zmniejszanie się wartości szerokości impulsu przez co serwomechanizm zacznie się cofać do pozycji początkowej. Czynność ta jest powtarzana przez 20 sekund. Po tym czasie następuje wyczyszczenie interwału czasowego co zatrzymuje działanie serwomechanizmu.
+\subsection{Sekwencyjne wykonywanie komend}
+W celu implementacji sekwencyjnego uruchamiania elementów podłączonych do pinów GPIO mikrokontrolera zastosowano asynchroniczne funkcje. Sekwencyjne wywoływanie funkcji to sposób, który umożliwi wykonywanie zaimplementowanych metod po kolei. Kiedy jedna funkcja zostanie zakończona, następuje wywołanie kolejnej. Zastosowanie asynchronicznego mechanizmu pozwoli na zaimplementowanie funkcji, które będą się wykonywały jedna po drugiej. Kiedy pierwsza operacja zrealizuje swoją pracę, informacja o jej zakończeniu zostanie przekazana do kolejnej funkcji, która następnie wykona swoją pracę. Pozwoli to na osiągnięcie efektu sekwencyjnego wykonywania się poleceń. 
+\newpage
+\begin{lstlisting}[caption=Sekwencyjne uruchamianie się funkcji]
+async function sequential() {
+    await turnOnAll()
+    await turnOffAll()
+    await tryRed()
+    await endBlinkRed()
+    await tryGreen()
+    await endBlinkGreen()
+    await tryYellow()
+    await endBlinkYellow()
+    await flowingLeds()
+    await blinkRed()
+    await blinkGreen()
+    await blinkYellow()
+    await servo()
+}
+\end{lstlisting}
+\section{Archiwizacja wpisanych poleceń}
+Archiwizacja wiadomości w aplikacjach opartych o komunikację tekstową jest używana, aby umożliwić użytkownikom przechowywanie i przeglądanie wiadomości, które zostały wysłane lub otrzymane w przeszłości. Jest to szczególnie przydatne, w przypadku osiągnięcia dostępu do ważnych informacji zawartych w konwersacji i możliwości nadzorowania danych zawartych w aplikacji. Archiwizacja pozwala także na lepsze zarządzanie dostępem do wiadomości i uniknięcie sytuacji, w której tracona jest zawartość istotnych danych. Możliwość zachowania archiwum wiadomości znajduje również swoje zastosowanie w zarządzaniu bazami danych, w dziale automatyki. W szczególności archiwizacja pozwala na zachowanie dostępu do poprzednich wersji danych, które zostały zmodyfikowane lub zaktualizowane. W niektórych przypadkach, możliwość zapisania informacji może być również ważna w kontekście utraty danych w wyniku awarii aplikacji lub innych problemów technicznych.
```

Comment:
Celem archiwizacji wiadomości jest możliwość przeglądania przez użytkownika historii komend wykonanych w aktualnej lub poprzedniej sesji. Z ich pomocą użytkownik wie, czy dana komenda została już wykonana, a jeżeli tak to kiedy oraz kto ja wykonał. Historia wiadomości może okazać się ważna w przypadku nieoczkiwanego przerwania pracy z systemem, jak na przykład restart komputera lub usunięcie karty z przeglądarki. 

W celu przygotowania archiwizacji, stworzono plik o formacie .JSON, który przechowuje kolejne wpisane komendy użytkownika. Wiadomości nie muszą zawierać komend do wykonania przez maszynę. Mogą być komunikatem do innego użytkownika znajdującego się w pokoju. Wykorzystanie pliku JSON zamiast bazy danych pozwoliło uniknąć dodatkowej złożoności aplikacji. Jednocześnie, operacje wykonywane na pliku są równie szybkie jak te bazodanowe. Same zarządzanie plikiem nie wymaga instalacji dodatkowych bibliotek, a jest możliwe dzięki wykorzystanie modułów dostępnych w środowisku uruchomieniowym Node.js. Dodatkowo, aplikacja nie wymaga posiadania serwera bazodanowego, co stanowczo ułatwia tworzenia jak i zarządzanie aplikacją.